### ISSUE TITLE
hang on 3.7 with bash loop

### ISSUE THREAD

**Issue by jumanjiman** [gliderlabs/docker-alpine#363]
Given this script:

```
#!/bin/bash

# https://github.com/koalaman/shellcheck/wiki/SC2044
while IFS= read -r -d '' file; do
  echo -n checking "${file}"

  # This works on alpine 3.6 but hangs on 3.7 and edge.
  if stat -c %F "${file}" | grep -qE '^directory$'; then
    echo " DIRECTORY"
  else
    echo
  fi

done < <(find /etc -maxdepth 1 -print0)
```

It works on alpine 3.6.

On alpine 3.7 and edge, it hangs after printing the first filename.

Reported upstream at https://bugs.alpinelinux.org/issues/8236 

____

**update**: additional reproducers

```
cat <(ls)
```

```
bash -c '(:) & wait %1'
```

**Comment by jumanjiman** [comment]
Assume the above script is named `reproducer` and is executable.

This Dockerfile:

```
FROM alpine:3.7
RUN apk add --no-cache bash
COPY reproducer /usr/bin/
```

Build the image:

```
$ docker build --rm -t reproducer .
Sending build context to Docker daemon  3.072kB
Step 1/3 : FROM alpine:3.7
 ---> e21c333399e0
Step 2/3 : RUN apk add --no-cache bash
 ---> Running in da7ef8c46fdf
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/x86_64/APKINDEX.tar.gz
(1/6) Installing pkgconf (1.3.10-r0)
(2/6) Installing ncurses-terminfo-base (6.0_p20170930-r0)
(3/6) Installing ncurses-terminfo (6.0_p20170930-r0)
(4/6) Installing ncurses-libs (6.0_p20170930-r0)
(5/6) Installing readline (7.0.003-r0)
(6/6) Installing bash (4.4.12-r1)
Executing bash-4.4.12-r1.post-install
Executing busybox-1.27.2-r6.trigger
OK: 13 MiB in 17 packages
 ---> d35e832604d3
Removing intermediate container da7ef8c46fdf
Step 3/3 : COPY reproducer /usr/bin/
 ---> 4ce6b6f753f3
Successfully built 4ce6b6f753f3
Successfully tagged reproducer:latest
```

Run it:

```
$ docker run --rm -it reproducer
/ # /usr/bin/reproducer 
checking /etc^C
/ #
```

Change the Dockerfile to `FROM alpine:3.6`, build, and run:

```
$ docker run --rm -it reproducer
/ # /usr/bin/reproducer 
checking /etc DIRECTORY
checking /etc/protocols
checking /etc/securetty
checking /etc/TZ
checking /etc/conf.d DIRECTORY
checking /etc/profile.d DIRECTORY
checking /etc/opt DIRECTORY
-snip-
checking /etc/modules
checking /etc/modprobe.d DIRECTORY
checking /etc/resolv.conf
/ # 
```

**Comment by tianon** [comment]
Oh man, this sounds very similar to https://github.com/tianon/docker-bash/issues/4, which I've now replicated on Alpine 3.7's `bash` package!

So, a simpler reproducer is to simply run `cat <(ls)` in bash, and it'll hang.

**Comment by jumanjiman** [comment]
Bizarrely, this issue does not occur on https://labs.play-with-docker.com/

IOW, on play-with-docker start an alpine:edge container and attempt to reproduce.
None of the reproducers hang when using play-with-docker.

However, the reproducers *do hang* in my local dev environments.

This seems to rule out process substitution by itself.
Perhaps it's some combination of tty handling and process substitution.

**Comment by ncopa** [comment]
should be fixed with `bash-4.4.12-r2` in v3.7.

**Comment by jumanjiman** [comment]
> should be fixed with bash-4.4.12-r2 in v3.7.

confirmed.
